### ISSUE TITLE
TagUI solved an issue that had been months unresolved

### ISSUE THREAD

**Issue by ronmoody** [kelaberetiv/TagUI#485]
Thank you, Ken, for your making this tool available to the rest of us. Until last week, I hadn't been able to automate 40 still shots in a Panasonic Pan Tilt Zoom (PTZ) camera. Neither the company nor any related/recommended source had provided anything close to a solution. But in one day, despite not being a programmer, I was able with TagUI to generate a script that actually worked. Subsequently, I revised that script into four versions and used task scheduler to even automate the timing of the scripts. It really worked. What took me hours on a weekly basis now doesn't even take seconds.

So thank you. The only thing I haven't figured out is how to insert date and time into the filename rather than creating an arbitrary filename so the saved files don't overwrite each other. The current command is "snap page to ..desktopattendanceWed7pm03.jpg", but I'd rather have it look like yymmddhhss.jpg. Right now, for example, would look like 190704074640.jpg (2019, July, 04, 07:46:40am in jpeg format. This would allow me to use one script rather than four, and those jpg files would never overwrite.

Again, I'm so very grateful. Thank you from Maui, Hawaii, USA
Ron Moody

**Comment by kensoh** [comment]
Hi Ron, thanks for sharing your nice experience with using TagUI! I appreciate it.

Credit go to AI Singapore 👍 TagUI is a project that I developed while on a sabbatical to Europe in 2016-2017. Then in 2018, I joined AI Singapore to further develop it. I left later that year after publishing a release which I felt is pretty complete and stable. AI Singapore continued maintaining it while I make time to contribute when I can. I'm encouraged and impressed by the automation script you created 👏 

The easiest way to achieve that is to put a tagui_local.js file in the same folder as your script. Below would be the content of tagui_local.js. TagUI allows defining [custom JavaScript functions](https://github.com/kelaberetiv/TagUI#helper-functions) to do more.

**tagui_local.js**
```js
function pad2(n) { // return 2-character string
    return (n < 10 ? '0' : '') + n;
}

function timestamp() { // return YYMMDDHHMMSS string
    var d = new Date();
    return d.getFullYear().toString().substr(-2) +
        pad2(d.getMonth() + 1) +
        pad2(d.getDate()) +
        pad2(d.getHours()) +
        pad2(d.getMinutes()) +
        pad2(d.getSeconds());
}
```

After that, using below in your script would save the image with appropriate filename (\`variable\` is the syntax in TagUI for using JavaScript variables or functions in place of text). Have fun! -

```js
snap page to `timestamp()`.jpg
```

Below are references used to get above JavaScript solution - 
- https://www.w3schools.com/js/js_date_methods.asp
- https://jsfiddle.net/felipekm/MYpQ9/
- https://stackoverflow.com/questions/17306830/how-to-get-2-digit-year-w-javascript

**Comment by kensoh** [comment]
Closing the issue for now, but do post here if there is new inputs!

**Comment by ronmoody** [comment]
Thank you Ken for your still-working fix on TagUI to automate photos taken
with our PTZ camera.

I come to you today with an unrelated request.

Can you give me a recommendation for an AI project - open source if
possible?

Here's the essence. I have an MSSQL database that has loads of church
information, but it's extremely hard to use; to get info in-to and out-of.
My goal is to remove blocks between leaders and the information inside my
MSSQL database. For example, if someone on Maui texts a prayer request via
our Twilio number, I want to use their phone number to see if they're in a
home group and send the leader their prayer request. If they're not in a
group I want to send a copy of their request to the leader responsible for
people living in the area where they live (area coordinator) so someone can
pray personally with them. Another example - I'd like to text a group
leader or area coordinator names and phone numbers for people with a
birthday today so they can call them and wish them a happy birthday. There
are thousands of other examples but I'll leave you with these two.

In other words, I want to use the database info to leverage human
encouragement and support.
In July of 2019 you helped me automate a pan-tilt-zoom camera to take 40
stills of people sitting in our church service (every service). We don't
use these photos for automated facial recognition. Instead, a group of us
reviews the photos every Tuesday morning, noting who is in church and who
isn't. We then use the notes we just took to help us follow-up. Again,
using technology to enable human interaction and support.

You helped me before. Can you help me again?
Ron Moody
King's Maui
***@***.***

**Comment by kensoh** [comment]
Hi Ron, oh wow it has been ages! Glad to know that TagUI has been useful to your church-related process.

I've since left AI Singapore and now @ruthtoh has taken over the project. And I've moved to London, UK.

From your description, you can try out these 2 options below. Both of them feature integrations with MySQL and Twilio. And both are designed to be very user-friendly. I have not personally used them before, but I can see their value proposition to customers who wants to easily automate certain apps that are in their list of supported apps.

With them, you probably could use MySQL data as input, and then use Twilio as input to receive message events to trigger a process, before finally using Twilio or email as output to notify the group leader. You'll probably have to try out both, to see which ones meet your requirements. They might have free tiers if the volume of work done is not too much.

Using open-source from scratch using programming languages is too much work. But if you have the bandwidth, you could check out Node-RED. It would be, in my opinion, the closest to what would be an open-source version of Zapier or Airtable. There is also n8n but it is not real open-source and free, but only using open-source for marketing and easier distribution, in my limited opinion.

**Comment by ronmoody** [comment]
Thank you for your insight as well as your suggestions. Would the same be
true of MSSQL, which is the database I'm trying to integrate?
ron

**Comment by kensoh** [comment]
Yes I believe all the above tools integrate with MySQL.
